File: Java version/app/src/main/java/it/unical/mat/parsers/pddl/solver_planning_domains/SPDGrammarBaseVisitorImplementation.java
package it.unical.mat.parsers.pddl.solver_planning_domains;

import it.unical.mat.parsers.pddl.PDDLDataCollection;
import org.antlr.v4.runtime.BailErrorStrategy;
import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.ConsoleErrorListener;
import org.antlr.v4.runtime.DefaultErrorStrategy;
import org.antlr.v4.runtime.RecognitionException;
import org.antlr.v4.runtime.atn.PredictionMode;

public class SPDGrammarBaseVisitorImplementation extends SPDGrammarBaseVisitor <Void> {
	private static final int OK_STATUS = 1;
    private static final int ERROR_STATUS = -1;
    private int status;
    private final PDDLDataCollection actions;
    private String errors = "";

    private SPDGrammarBaseVisitorImplementation(final PDDLDataCollection actions) {
        this.actions = actions;
    }

    @Override
    public Void visitPair(SPDGrammarParser.PairContext ctx) {
        final SPDGrammarParser.ValueContext valueContext = ctx.value();
        final String string = ctx.STRING().getText();

        if(status == 0 && string.equalsIgnoreCase("\"status\""))
            status = valueContext.getText().equalsIgnoreCase("\"ok\"") ? OK_STATUS : ERROR_STATUS;
        else if(status == ERROR_STATUS) {
            if(string.equalsIgnoreCase("\"result\"")) {
                if(valueContext instanceof SPDGrammarParser.ArrayValueContext || valueContext instanceof SPDGrammarParser.ObjectValueContext)
                    return visitChildren(ctx);
                else
                    errors += trim(valueContext.getText());
            } else if(string.equalsIgnoreCase("\"error\""))
                errors += trim(valueContext.getText());
        } else if(status == OK_STATUS) {
            if(string.equalsIgnoreCase("\"name\""))
                actions.storeAction(trim(valueContext.getText()));
            else if(string.equalsIgnoreCase("\"plan\"") || string.equalsIgnoreCase("\"result\""))
                return visitChildren(ctx);
        }

        return null;
    }

    private static String trim(final String string) {
        final int stringLength = string.length();

        return (string.charAt(0) == '"' && string.charAt(stringLength - 1) == '"') ? string.substring(1, stringLength - 1) : string;
    }

    public static String parse(final PDDLDataCollection actions, final String spdOutput, final boolean two_stageParsing) {
    	final CommonTokenStream tokens = new CommonTokenStream(new SPDGrammarLexer(CharStreams.fromString(spdOutput)));
        final SPDGrammarParser parser = new SPDGrammarParser(tokens);
        final SPDGrammarBaseVisitorImplementation visitor = new SPDGrammarBaseVisitorImplementation(actions);
        
        if(!two_stageParsing) {
        	visitor.visit(parser.json());
        	
        	return visitor.errors;
        }
        
        parser.getInterpreter().setPredictionMode(PredictionMode.SLL); 
        parser.removeErrorListeners();
        parser.setErrorHandler(new BailErrorStrategy());
     
        try {
        	visitor.visit(parser.json());
        } catch (final RuntimeException exception) {
        	if(exception.getClass() == RuntimeException.class && exception.getCause() instanceof RecognitionException) {
        		tokens.seek(0);
        		parser.addErrorListener(ConsoleErrorListener.INSTANCE);
        		parser.setErrorHandler(new DefaultErrorStrategy());
        		parser.getInterpreter().setPredictionMode(PredictionMode.LL); 
                visitor.visit(parser.json());
        	}
        }

        return visitor.errors;
    }
}

File: CSharp version/it/unical/mat/parsers/pddl/solver_planning_domains/SPDGrammarBaseVisitorImplementation.cs
using System;
using Antlr4.Runtime;
using Antlr4.Runtime.Atn;

namespace it.unical.mat.parsers.pddl.solver_planning_domains
{
    public class SPDGrammarBaseVisitorImplementation : SPDGrammarBaseVisitor<object>
    {
        private const int OK_STATUS = 1;
        private const int ERROR_STATUS = -1;
        private int status;
        private readonly IPDDLDataCollection actions;
        private string errors = "";

        private SPDGrammarBaseVisitorImplementation(IPDDLDataCollection actions)
        {
            this.actions = actions;
        }

        public override object VisitPair(SPDGrammarParser.PairContext context)
        {
            SPDGrammarParser.ValueContext valueContext = context.value();
            string @string = context.STRING().GetText();

            if (status == 0 && @string.Equals("\"status\"", StringComparison.CurrentCultureIgnoreCase))
                status = valueContext.GetText().Equals("\"ok\"", StringComparison.CurrentCultureIgnoreCase) ? OK_STATUS : ERROR_STATUS;
            else if (status == ERROR_STATUS)
            {
                if (@string.Equals("\"result\"", StringComparison.CurrentCultureIgnoreCase))
                {
                    if (valueContext is SPDGrammarParser.ArrayValueContext || valueContext is SPDGrammarParser.ObjectValueContext)
                        return VisitChildren(context);
                    else
                        errors += Trim(valueContext.GetText());
                }
                else if (@string.Equals("\"error\"", StringComparison.CurrentCultureIgnoreCase))
                    errors += Trim(valueContext.GetText());
            }
            else if (status == OK_STATUS)
            {
                if (@string.Equals("\"name\"", StringComparison.CurrentCultureIgnoreCase))
                    actions.StoreAction(Trim(valueContext.GetText()));
                else if (@string.Equals("\"plan\"", StringComparison.CurrentCultureIgnoreCase) || @string.Equals("\"result\"", StringComparison.CurrentCultureIgnoreCase))
                    return VisitChildren(context);
            }

            return null;
        }

        private static string Trim(string @string)
        {
            return (@string[0] == '"' && @string[@string.Length - 1] == '"') ? @string.Substring(1, @string.Length - 2) : @string;
        }

        public static string Parse(IPDDLDataCollection actions, string spdOutput, bool two_stageParsing)
        {
            CommonTokenStream tokens = new CommonTokenStream(new SPDGrammarLexer(CharStreams.fromstring(spdOutput)));
            SPDGrammarParser parser = new SPDGrammarParser(tokens);
            SPDGrammarBaseVisitorImplementation visitor = new SPDGrammarBaseVisitorImplementation(actions);

            if (!two_stageParsing)
            {
                visitor.Visit(parser.json());

                return visitor.errors;
            }

            parser.Interpreter.PredictionMode = PredictionMode.SLL;

            parser.RemoveErrorListeners();

            parser.ErrorHandler = new BailErrorStrategy();

            try
            {
                visitor.Visit(parser.json());
            }
            catch (SystemException exception)
            {
                if (exception.GetBaseException() is RecognitionException)
                {
                    tokens.Seek(0);
                    parser.AddErrorListener(ConsoleErrorListener<object>.Instance);

                    parser.ErrorHandler = new DefaultErrorStrategy();
                    parser.Interpreter.PredictionMode = PredictionMode.LL;

                    visitor.Visit(parser.json());
                }
            }

            return visitor.errors;
        }
    }
}